src/analysis/vector/geometry_checker/qgsfeaturepool.cpp

```diff
@@ -47,6 +48,17 @@ bool QgsFeaturePool::getFeature( QgsFeatureId id, QgsFeature &feature )
   //
   // https://bugreports.qt.io/browse/QTBUG-19794
 
+  // a memory layer is already a cache itself. retrieve feature from the feature source.
+  if ( mIsMemoryLayer )
+  {
+    if ( !mLayer->getFeatures( QgsFeatureRequest( id ) ).nextFeature( feature ) )
```

Comment:
This isn't thread safe -- eg. in the geometry checker plugin the feature pool is used from a background thread after constructing the source (which IS a thread safe approach, the feature source was created on the main thread), but we can't just swap this out to access the layer here because accessing the **layer** from a different thread is not thread safe.

Reply:
Right, I see the difference between "being in another thread and accessing the feature source created on the main thread" and "being in another thread and accessing the layer created on the main thread".

It is complicated, today the geometry checker only works because updating the feature source **does** indeed update the underlying data (we discussed this in #58113)... which is not the case with a memory layer...

In fact, the question is "how do I get a feature that is up-to-date for a memory layer that lies in another thread?"

What are your thoughts about:
- option 1: encapsulate this not-thread-safe statement in a lambda with `QgsThreadingUtils::runOnMainThread()`
- option 2: have a check that the layer is in the correct thread and return `false` or handle an exception?
- option 3: make `getFeatures()` for a memory layer thread-safe, with mutex etc. ?

Reply:
@nyalldawson this issue is very important, because the PR for exposing all geometry checker algorithms to the processing toolbox depends on it (the code is ready, I could open the PR, but it relies on this commit at the moment).

Reply:
@Djedouas 

Honestly, I don't particularly like any of those 3 options. Especially not 3. Don't do 3. Run from that one :rofl: 

I think trying to patch around this in the QgsFeaturePool class like this might be the wrong approach. My understanding of the issue is that the geometry checker is changing features in the layer, and then the feature cached in the pool is out of date with that. If so, then a couple of things jump to mind:

1. Since the geometry checker has knowledge of the feature pool already, what about adding an explicit "set updated feature" method to the pool? There could be a QHash< QgsFeatureId, QgsFeature > to handle this, so when "setUpdatedFeature" is called then the feature is added to this hash. Then getFeature would first check if the feature is the updated feature hash, and if so, it just returns that feature instead of querying the feature source. Admittedly it's not the most elegant solution, but it would be a pragmatic fix for the problem which doesn't have any risk of deadlocks (as runOnMainThread would have), and doesn't involve messing with any of the fundamental, low-level threading design around the feature sources/vector layers (as adding the mutex to the memory provider would do).
2. I think there's potentially a deeper issue at play here too :scream: . Are you updating features in the vector layer from a background thread? (eg in a processing algorithm's `processAlgorithm` function). If so, that's not thread safe, and should NOT be done. You instead need to queue up the changes for the layer, and then apply them ONLY in the postProcessAlgorithm step, which runs in the main thread after processAlgorithm finishes.

Reply:
@nyalldawson thanks, see my thoughts below

About point 1.

Your point made me think of another way to fix the specific memory layer issue. You're right, we have a cache that is out of date with the real geometry, and the geometry checker gets the feature from the cache when fixing geometry checks. But I thought that, other than creating yet another sub-cache with up-do-date features, I could take a look at the method refreshing the cache.

So here is a new commit / new way to fix the problem: in `refreshCache` (which was using `getFeature`, and falling in the no-cache branch there, and getting the out-of-date geometry from the feature source) we can directly set the up-to-date feature and its geometry.

About point 2.

In all incoming processing algorithms that will use the geometry checker and feature pool logic, a copy of the input layer is created within the `processAlgorithm` function, I don't use the multi-threading code of the current geometry checker plugin, and everything checking/modifying the layer is done in the processing thread, where the layer is from. We can discuss this further on in the dedicated PR I will submit.